Add autocomplete for emoji, users, chans, and commands

yashsriv opened the pull request:
Problems currently being faced:
* js linting error for undefined `EmojiConvertor`: Any suggestions on how to fix this
* Pressing Up and Down keys to cycle through autocompletion, also cycles through history. Someway to disable the history plugin or prevent key events from going ahead while autocomplete is active would be nice.

![preview](http://i.imgur.com/RMkndQX.png)

yashsriv commented:
I was also thinking that we could improve the results by using https://github.com/andreruffert/emoji-suggestions which is a better search - you can search food to get all food related emojis and so on.

yashsriv commented:
bump. I think this is ready, if there are no better solutions possible for disabling the history plugin.

MaxLeiter commented:
Maybe we wait for #640 so we can use npm?

xPaw commented:
I think if we are going to implement auto completion, it should be done along with nick auto completion.

YaManicKill commented:
> if we are going to implement auto completion

First of all, I think we need to do this at some point. Whether it's this one or another, we can't say no to this in general. It is the third most :+1: feature request.

> it should be done along with nick auto completion

If you mean that a PR should be added with both, I disagree. Let's add 1 and then the other. There's no need to tie the 2 together.

If you mean this should be done in such a way to make nick completion incredibly easy, then yes.

I'll take a look at the PR soon.

xPaw commented:
> There's no need to tie the 2 together.

Are you sure? It would most likely end up with a cleaner over all solution if they were strongly tied in.

> If you mean this should be done in such a way to make nick completion incredibly easy, then yes.

I mean to allow displaying matched names in a dropdown for selection, just as emoji. This would also help on mobile.

yashsriv commented:
@xPaw Currently for nick autocompletion, we manually trigger it with tab, while this is triggered automatically. But I think its possible to combine both for a cleaner solution as you suggest - since pressing tab on mobile is not possible. I'll try working on it tomorrow. 👍 Thanks for the suggestion.

yashsriv commented:
Added nickname auto completion as well. Should I add an option for enabling nick complete / emoji complete / both?? The nick complete is slightly invasive imo.

* Emoji completion is essentially triggered by a word starting with `:`
* Nickname completion is triggered after 2 letters are input - We could modify it such that it is triggered by a word starting with `@` - but once the user chooses to autocomplete, `@` is removed. Something similar happens for emoji where the `:` and subsequent characters are replaced by the unicode emoji

Cldfire commented:
> Nickname completion is triggered after 2 letters are input - We could modify it such that it is triggered by a word starting with @ - but once the user chooses to autocomplete, @ is removed.

Typing @ to initiate nick autocomplete is surely the better way to go, most people (including myself) won't want to have the nick autocomplete window popping up every single word :)

YaManicKill commented:
Yeah, can you enable nickname auto completion with @ symbol please ☺ great
work.

On Tue, 20 Dec 2016, 03:35 Cldfire, <notifications@github.com> wrote:

> Nickname completion is triggered after 2 letters are input - We could
> modify it such that it is triggered by a word starting with @ - but once
> the user chooses to autocomplete, @ is removed.
>
> Typing @ to initiate nick autocomplete is surely the better way to go,
> most people (including myself) won't want to have the nick autocomplete
> window popping up every single word :)
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/thelounge/lounge/pull/787#issuecomment-268145911>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACD_4f_-uZLWpF5haBFAddDoQgyWFakyks5rJ00BgaJpZM4LJzGv>
> .
>

astorije commented:
This is really awesome, great stuff @yashsriv! I haven't tried it yet, will do after @xPaw or @YaManicKill start giving their 👍.

I like a lot the idea of typing `@` to trigger the nick complete then removing it when selecting. If it turns out using `@nick` is common, we could do it with 2 `@`: first one gets removed, second one does not and does not influence the search results. Maybe stupid, but just throwing an idea :)

Regarding style, what about binding the dropup above the input and making it full-width? Something like the following (screencast from Telegram):

![telegram_emoji](https://cloud.githubusercontent.com/assets/113730/21340249/49b11890-c653-11e6-8ae2-93ed319421f2.gif)

xPaw commented:
Just going to throw this in, this dropdown could also autocomplete IRC colors after inserting a `\x03` (using ctrl+k hotkey), and would suggest all the colours (00-15).

YaManicKill commented:
> Just going to throw this in, this dropdown could also autocomplete IRC colors after inserting a \x03 (using ctrl+k hotkey), and would suggest all the colours (00-15).

Good idea, there are a few other things we can do as well (channels on #, etc)

But let's **definitely not** include this in the PR. I'll take a look at the code soon.

yashsriv commented:
All right. This is it. I promise I won't make anymore major changes to the code 😝 @YaManicKill .
Four completion strategies:

* `emoji`: Triggered by `:` and as described earlier
* `nicks`: Triggered by `@` and 1 other letter - `@` is removed on completion.. `@@` allows `@` to remain
* `commands`: Triggered by `/` at the beginning of the line
* `chans`: Triggered by the beginning of a valid chan name as per rfc standard

astorije commented:
What do we autocomplete chans with? Chans being currently joined?

Also, what do people think of my UI suggestion above? Yes, no? :) It doesn't have to go with this PR obviously, but I'm curious. I think it would look more integrated.

yashsriv commented:
@astorije Yup. Chan names are completed from all channels in the sidebar

And I think the full width dropup might look better - I think this one doesn't look so good on mobile.

YaManicKill commented:
> Also, what do people think of my UI suggestion above? Yes, no?

Possibly, but I think I'd want to play with both of them. I'm not convinced about all the whitespace, but we can see.

Shall we get this one in first, and then we can play around with different looks afterwards? Probably an easier way to find a good way for it to look.

YaManicKill commented on client/css/bootstrap.css:
```diff
@@ -1022,6 +1022,127 @@ tbody.collapse.in {
   -o-transition: height 0.35s ease;
   transition: height 0.35s ease;
 }
+.dropdown {
```
I really want to be pedantic and point out that it's actually a drop-up, not a dropdown.

yashsriv replied:
I've actually copied the styling entirely from bootstrap - This particular css exists in extended bootstrap with the dropdown included. That's why I appended this to bootstrap.css and not generic css. Plus I think the div that is added has the class `.dropdown` by default (customizable but default is .`dropdown` for compatibility with extended bootstrap.css)

YaManicKill replied:
I was just being pedantic, I don't actually think you should rename it :-P

YaManicKill commented on client/js/libs/jquery/inputhistory.js:
```diff
@@ -54,7 +54,7 @@
 			case 40: // Down
 				// NOTICE: This is specific to The Lounge.
-				if (e.ctrlKey || e.metaKey) {
+				if (e.ctrlKey || e.metaKey || self.data("autocompleting")) {
```
I'm not massively worried about this, because the library was built for shout originally anyway. I'm not sure of a better way of doing this, and we've modified it loads anyway.

YaManicKill commented on client/js/lounge.js:
```diff
@@ -1308,14 +1402,31 @@ $(function() {
+	function completeNicks(word) {
 		var users = chat.find(".active").find(".users");
-		var nicks = users.data("nicks");
+		var words = users.data("nicks");
```
This would fix #64 as well, which is good.

YaManicKill commented on client/js/lounge.js:
```diff
@@ -52,6 +52,84 @@ $(function() {
+
+	var nicksStrategy = {
+		id: "nicks",
+		match: /\B@([a-zA-Z_[\]\\^{}|`@][a-zA-Z0-9_[\]\\^{}|`-]*)$/,
```
I think we need to still allow tabcomplete without the @ (so this should work as it currently does, but if I just type "ya<tab>" it should autocomplete with "yamanickill" as that's how a lot of people use it.

xPaw replied:
Until we can auto complete nicks the same way we currently can, IMO this PR can't go in. It's an absolutely breaking change that would be annoying to deal with.

yashsriv replied:
If we still need the older behavior of tabcomplete, I can reintroduce the earlier plugin and enable `completeNicks` for it.

YaManicKill replied:
@yashsriv I don't want the autocomplete popup to appear when just typing names. The old feature needs to work exactly the same as before.

2 use cases:

1. Tab complete. Work exactly as before.

Type characters, press <tab>, get autofilled with rest of username. No popup, no nothing.

2. "@", work exactly as you have it here.

Press "@" and a letter, get popup with list of usernames, gets filtered as you type letters.

The old one you had, where the autocomplete would pop up when you just type a few letters of a username was far too intrusive and would be horrible.

As I said in my overall comment, if you can't figure out a nice way to keep both of these use cases, just remove the username autocomplete from this PR and re-enable tab-complete, and we'll figure it out in another PR. I'd rather another PR than this one takes months to get in. Your work is pretty good and I would love to get this in sooner rather than later. Let's not feature-creep it. We've got a good structure for adding these, which is what @xPaw wanted when he suggested adding users to this.

yashsriv replied:
The latest commits do precisely this
>2 use cases:
>
>    1. Tab complete. Work exactly as before.
>
>Type characters, press , get autofilled with rest of username. No popup, no nothing.
>
>   2. "@", work exactly as you have it here.

YaManicKill replied:
Great, will just test it out before approving :-)

YaManicKill commented on client/js/lounge.js:
```diff
@@ -52,6 +52,84 @@ $(function() {
+
+	var chanStrategy = {
+		id: "chans",
+		match: /\B(#|\+|&|![A-Z0-9]{5})([^\x00\x0A\x0D\x20\x2C\x3A]+(:[^\x00\x0A\x0D\x20\x2C\x3A]*)?)$/,
```
This appears to only filter based on the first letter. If I have 2 channels called "#pokemon" and "#pkmn", and I type "#po" then "#pkmn" will still be at the top of the list.

yashsriv replied:
Fixed this 😃

astorije commented:
> Shall we get this one in first, and then we can play around with different looks afterwards? Probably an easier way to find a good way for it to look.

Yep, sounds good to me!

I'll try to review carefully this weekend :)

MaxLeiter commented:
I think we should mimic slack in overriding `enter` to work as tab-complete when the suggestion modal is up. (meaning enter should act as tab when auto-complete menu is showing)

astorije commented:
Alright, I played a bit with this, and I'm impressed, it's really cool and will definitely add quite a lot to the project. v2.2.0 is around the corner, but I really think this could/should go with the release after that, v2.3.0.

Before we get there, I've noticed a few things that needs updating though:

- Now that #640 is out, you can use the [npm package](https://www.npmjs.com/package/jquery-textcomplete) directly and remove the library from this PR. That will also get rid of your linting issues!
- I agree with @MaxLeiter above, hitting <kbd>Enter</kbd> should select the currently selected entry. Heck, I've sent like 15 messages instead of picking an emoji/channel/nick while testing this lol.
- Style of your dropdowns do not match the current ones (contextual menu on nicks, channels, top-right menu button), could you match/reuse it please?
- Completion dropdown opens as soon as `:` is pressed but for `#` and `@` it needs an extra character. They should all open as soon as one of these "trigger" characters is entered.
- Selecting an emoji adds a space while selecting a channel or nick do not. Selecting an emoji should not add a space to be consistent with the other (and match the current behavior of nick completion).

Once that PR lands, I'll remove the logic introduced by #708 as it's redundant and your result is unarguably better. People who hit this button instead of Send will be happy 😅.

> Until we can auto complete nicks the same way we currently can, IMO this PR can't go in. It's an absolutely breaking change that would be annoying to deal with.

@xPaw, what do you mean by "Until we can auto complete nicks the same way we currently can"? I believe it's possible, granted you need to type `@` first. It would be a shame to have 2 libraries for the same thing really.
If you don't want to have to type `@`, Slack is solving this by opening the nick suggestions when hitting <kbd>Tab</kbd> on any text (so dropdown opens as soon as typing `@` or when typing characters then hitting <kbd>Tab</kbd>). @yashsriv, do you know if it is possible to do that with the library you are using?

yashsriv commented:
I have updated the UX as you specified @astorije :
- [x] Hitting <kbd>ENTER</kbd> now completes the completion
- [x] Used npm package for jquery textcomplete
- [x] The dropdown opens instantly on pressing `@` or `#`
- [x] Selecting Emoji should not insert space
Tasks left:
- [ ] Match style of dropdowns
- [ ] Manually trigger completion with <kbd>Tab</kbd>

yashsriv commented:
I have updated the UX as you specified @astorije :
- [x] Hitting <kbd>ENTER</kbd> now completes the completion
- [x] Used npm package for jquery textcomplete
- [x] The dropdown opens instantly on pressing `@` or `#`
- [x] Selecting Emoji should not insert space
Tasks left:
- [ ] Match style of dropdowns
- [ ] Manually trigger completion with <kbd>Tab</kbd>

Also, is there anyway to not include `emoji.js` while linting

astorije commented:
> Also, is there anyway to not include emoji.js while linting. (Found it)

We should not ship this library in the repo. Problem is, it's not on npm (see https://github.com/iamcal/js-emoji/issues/28).

Can you remind me why we need this?

yashsriv commented:
This library provides the map from word to emoji, i.e. `:smiley:` to 😃

yashsriv commented:
Found this: https://www.npmjs.com/package/emoji-name-map

Will try using it

astorije commented:
So if anyone (using The Lounge or not) posts a message on a channel that
contains " :smiley: ", users of The Lounge will see that rendered as an
emoji instead of text, correct?

If so, I think I'd rather stick with Unicode only. At least for this PR,
especially if we can't have the lib downloaded from npm. But more
generally, is this something we want? I'm not so sure, so I'd like to give
it a try with Unicode only and see how it goes. Thoughts,
@thelounge/maintainers?

yashsriv commented:
Uh no @astorije ... that was my earlier pull request and that was rejected by the lounge community. 😛 . I am using the earlier library to convert `smiley` to unicode equivalent. I think the new library can easily do this.

astorije commented:
This is really cool, can't wait to ship it!

A few remaining things I've noticed:

- [ ] Match style of dropdowns (but you're already aware of it
- [ ] `:rofl:` shows up as an empty square for me, maybe other emojis are doing that, is there anything we need to do about this?
  <img width="186" alt="screen shot 2017-02-12 at 17 25 17" src="https://cloud.githubusercontent.com/assets/113730/22866671/4180d456-f148-11e6-9b5a-7a1af6eb49cc.png">

I've noticed a few other things but they should be addressed when you fix the first item. Also, there seems to be some incompatibilities between our other context menus, but nothing super important or blocking, we can address this later.

YaManicKill commented:
@astorije The missing emoji will be because of the emoji on your system. Rofl is a unicode 9 emoji, meaning that it isn't in most things (latest iOS, latest Android is about it). I don't think we include any stylesheet for emoji specifically.

So we could include a specific stylesheet for emoji (Emoji one would probably be the best, as that is CC) but that would mean requiring people to have a specific emoji set rather than their system's set.

 ¯\_(ツ)_/¯

yashsriv commented:
Updated with matching context menu and rebased on master.

yashsriv commented:
@astorije Any more changes needed?? The only remaining thing I can think of is the tab completion. Currently, we have both the tab complete as well as the autocomplete popup.... Removing tab complete would be a breaking change for most users... So I'm not sure what we should do.

xPaw commented:
Tried latest changes, the auto completion is certainly much better than the last time I tried.

>  The only remaining thing I can think of is the tab completion. Currently, we have both the tab complete as well as the autocomplete popup....

Indeed that's a problem. Perhaps you could map `tab` key to complete the top suggestion in the menu? This however will certainly need some trickery for nick completion (as your menu uses `@` to show names).

Why did you make a separate style `textcomplete` instead of re-using the existing styles for dropdown menus? It already had styles for hover and stuff. Right now, there's no hover/active style if you use arrow keys for example.

williamboman commented on package.json:
```diff
@@ -44,10 +44,13 @@
     "fs-extra": "2.0.0",
     "irc-framework": "2.5.5",
+    "jquery-textcomplete": "^1.8.0",
+    "json-loader": "^0.5.4",
```
Use fixed dependency versions?

xPaw replied:
Yep.

I also don't think `json-loader` is even needed here as `emoji-name-map` is used.

xPaw commented on package.json:
```diff
@@ -44,10 +44,13 @@
     "cheerio": "0.22.0",
     "colors": "1.1.2",
     "commander": "2.9.0",
+    "emoji-name-map": "^1.2.2",
```
This is pretty much a wrapper for `emojilib` package. Said package also includes a `simplemap.json` file, which could be a better choice to use, as it doesn't use the full emojis.json.

yashsriv commented:
> Why did you make a separate style textcomplete instead of re-using the existing styles for dropdown menus? It already had styles for hover and stuff. Right now, there's no hover/active style if you use arrow keys for example.

@xPaw , the current css expects the `<li>` elements to have a certain class which cannot be controlled by me(the `li`s are inserted by `jquery-text-complete`). I can only set the `ul` class. I'll fix the hover and active css.

xPaw commented:
You can still use shared ul class.

astorije commented:
@yashsriv, can you address @xPaw's latest comment as well as rebase on master and fix conflicts? :)

Hang in there, you're almost ready to make it to master! 😄

williamboman commented:
I ran this for a while. What's the purpose of showing autocomplete for emojis with `:syntax:` if it doesn't interpolate it to unicode emojis?

YaManicKill commented:
> What's the purpose of showing autocomplete for emojis with :syntax: if it doesn't interpolate it to unicode emojis?

Wait, doesn't it? I thought the replace on the emoji strategy should do that, no?

```
replace: function(value) {
  return value;
},

williamboman commented:
It didn't when I tested it last week, I might be wrong though

yashsriv commented:
Rebased to current master and fixed CSS by copying the existing css for context menu. I can't reuse the existing context menu because it is styled by an id tag whereas I only have control over the class of the dropdown.

astorije commented:
> What's the purpose of showing autocomplete for emojis with :syntax: if it doesn't interpolate it to unicode emojis?

Maybe I misunderstand your question, but this is my attempt at answering it.
but if you type `:smi`, the `:smile:` choice will show up. If you then press <kbd>Enter</kbd> or <kbd>Tab</kbd>, the emoji replaces your `:smi` text. You send your message and the emoji appears again there. This probably works as you expected.
However, if you type `:smile:` in full, when you send your message, `:smile:` is not changed to an emoji. This is probably the case you are talking about. I'd argue this is good, because if you want to send `foo:man:bar` for any reason, you are not surprised after sending it that it got converted.

IMO, we should always be predictable, i.e. send what the user wrote. If user did not use an emoji, we should not send one. And vice-versa. Open to suggestions of course.

> I can't reuse the existing context menu because it is styled by an id tag

@yashsriv, you can use multiple selectors for a set of attributes. The following could do, but I haven't tested much. Also, it makes the dropdown look exactly like the context menu, which I find different in the current state of the PR.

```diff
──────────────────────────────
modified: client/css/style.css
──────────────────────────────
@ style.css:1484 @ kbd {
    background: transparent;
 }

-#context-menu {
+#context-menu,
+.textcomplete-menu {
    position: absolute;
    list-style: none;
    margin: 0;
@ style.css:1504 @ kbd {
    background-color: rgba(0, 0, 0, .1);
 }

-.context-menu-item {
+.context-menu-item,
+.textcomplete-item {
    cursor: pointer;
    display: block;
    padding: 4px 8px;
@ style.css:1514 @ kbd {
    margin-bottom: 6px;
 }

-.context-menu-item:hover {
+.context-menu-item:hover,
+.textcomplete-item:hover {
    background-color: #f6f6f6;
 }

-.context-menu-item:before {
+.context-menu-item:before,
+.textcomplete-item:before {
    width: 20px;
    display: inline-block;
 }

-.textcomplete-menu {
-   position: absolute;
-   list-style: none;
-   margin: 0;
-   padding: 0;
-   min-width: 160px;
-   background-color: #fff;
-   box-shadow: 0 3px 12px rgba(0, 0, 0, .15);
-   border: 1px solid rgba(0, 0, 0, .15);
-   border-radius: 2px;
-}
-
-.textcomplete-item {
-   cursor: pointer;
-   border-bottom-color: rgba(0, 0, 0, .1);
-   border-bottom-style: solid;
-   border-bottom-width: 1px;
-   display: block;
-   padding: 10px 8px;
-   color: #333;
-}
-
-.textcomplete-item:hover {
-   background-color: #f6f6f6;
-}
-
 .textcomplete-menu .active {
    background-color: #f6f6f6;
 }

-.textcomplete-item:before {
-   width: 20px;
-   display: inline-block;
-}
-
 .textcomplete-item a {
    color: #333;
 }
```

yashsriv commented:
@astorije I have updated the styles as suggested and added an entry on the help page 😄

astorije commented:
Regarding the help window:

- I think Tab/Enter/Up/Down are very straightforward. They should be automatically discoverable and discovered so I don't think we need them there. Others are more ad-hoc so yeah, good idea!
- Instead of `<key> + Any Letter`, I'd prefer to have a line between the "Autocompletion" title and the list that says something like: `Start typing the following characters followed by any letters to trigger the autocompletion dropdown:` (not a native speaker so probably a better wording). That way you can keep the symbol only in the list.
- Use `<code>` instead of `<kbd>`, I think.
- `Beginning of a valid chan name` is a bit awkward. Would it be sufficient to do `#/$` instead?
- Repeat of `Autocompletion` in the description is redundant. Just "nicknames"/"channels"/"emoji"/"commands" should do the trick (maybe "see list of available commands below" for the "commands" one?).

This starts to be reeeeeally sweet!! 😄

dgw commented:
> * Use `<code>` instead of `<kbd>`, I think.

`<kbd>` is the tag to use for keyboard input, so why use `<code>` when it's not code?

> * `Beginning of a valid chan name` is a bit awkward. Would it be sufficient to do `#/$` instead?

Then you have to get into servers that support `&` channels and other less common stuff. There's probably better wording, but if giving an example first character I'd give `#` _only_.

astorije commented:
> `<kbd>` is the tag to use for keyboard input, so why use `<code>` when it's not code?

I hear you. My thinking was that in the current page `<kbd>` is used to denote a key to hit while `<code>` denotes commands to type in the message input. So the former is the input really, the key, not matter what it produces, while the latter is the text produced no matter how (keyboard, copy-paste, etc.).
In this present case, the completion triggers in response to *typing the text*, not necessarily to *hitting the key*.
This is probably overthinking, but I just wanted to explain why I mentioned suggested this. Happy to change my mind if it doesn't make sense.

> Then you have to get into servers that support & channels and other less common stuff. There's probably better wording, but if giving an example first character I'd give # only.

That's reasonable. I suggested both because I used to work on a network where `&` was a big part of daily usage, but maybe it was a specific case. If 99% people use `#` only, then sure!

xPaw commented on package.json:
```diff
@@ -44,10 +44,13 @@
     "fs-extra": "2.1.2",
     "irc-framework": "2.6.1",
+    "jquery-textcomplete": "1.8.0",
+    "json-loader": "0.5.4",
```
This isn't required or used.

yashsriv replied:
`emojilib` can't import without this since in emojilib codebase, there is a require of a json file.

xPaw replied:
Ahh.

xPaw commented:
So this PR adds about ~300Kb to the JS build. I think emojilib is a big contributor to that as it has *a lot* of unnecessary data. You also need to mark these new libraries as vendor in `webpack.config.js`.

Besides that, seems shippable.

astorije commented:
> So this PR adds about ~300Kb to the JS build. I think emojilib is a big contributor to that as it has a lot of unnecessary data.

Unnecessary how? Is there a way we can cherry-pick these?

Also, @yashsriv, can you resolve conflicts please? :)
So clooose!!

xPaw commented:
@astorije Basically we only need a `string -> emoji` map, but the lib provides a lot of info, like keywords, categories, etc. It all adds up.

yashsriv commented:
@xPaw I have shifted to a set of new libraries for emoji which take only 75kB of space in the vendor bundle. The only issue I have with them is that both libraries essentially provide a list of emojis and list of emojieywords respectively. Since both are from the same author, I expect a 1-1 direct mapping of indice and that's how I have implemented it.

xPaw commented:
@yashsriv This approach seems prone to breaking. Also looking at the list, it seems rather weird: https://github.com/Kikobeats/emojis-keywords/blob/7a384119e86d13ec98dd1e71742c52db2555a169/index.js#L2342

What if we simply copied over `simplemap.json` directly into our project from emojilib? Then we have full control of the supported emojis. Not like it's hard to maintain.

xPaw commented on client/js/lounge.js:
```diff
@@ -75,6 +78,80 @@ $(function() {
+			}));
+		},
+		template: function(value) {
+			return emojiList[emojiKeywords.indexOf(value)] + " " + value;
```
Here's a suggestion:
```js
return `<span class="emoji">${emojiList[emojiKeywords.indexOf(value)]}</span> ${value}`;
```
```css
.textcomplete-item .emoji {
	width: 16px;
	text-align: center;
	display: inline-block;
}
```

This would introduce a nice consistency, since not all emojis are exactly sized.

yashsriv commented:
So we are back to using no npm for the emojis 😛

YaManicKill commented:
@yashsriv Have you addressed all comments and ready for review again?

xPaw commented:
Okay, I'm happy to merge this. Rebase this PR to latest master and add this style:
```css
.textcomplete-item .emoji {
	margin-right: 8px;
}```

yashsriv commented:
Rebased. @xPaw I think text-complete allows for fetching from an ajax request. But, wouldn't that request be resent everytime someone wanted to complete something?

xPaw commented:
@yashsriv I think we would have to implement that ourselves. Once the client is fully loaded, execute a request for `simplemap.json`, and once loaded, register emoji strategy in the textcomplete. Lazy af.

yashsriv commented:
I have tried removing tabcomplete on my branch here: https://github.com/yashsriv/lounge/tree/wip/notabcomplete

^ The above works but the code is convoluted (see the last commit). Feel free to try and improve it as I won't be free for atleast a week.

astorije commented:
@xPaw, are you still ok with this or expecting further changes?

xPaw commented:
@astorije 🚢

metsjeesus commented:
Just a old idea, because we are doing autocomplete. What do you think of completing first word on autocomplete as username to be "username, " or "username: "? Maybe make it configurable somewhere?

YaManicKill commented:
> What do you think of completing first word on autocomplete as username to be "username, " or "username: "? Maybe make it configurable somewhere?

So, let's not bring this up in this ticket. This PR is basically done, and is a nice piece of functionality on it's own. That discussion has already happened, it can happen again, but it's separate to this PR. If you want to talk about that, feel free to open a ticket (or re-open the existing one). But I don't think this is the right place.

YaManicKill commented:
Ah boo, we've got conflicts. @yashsriv You got time to resolve conflicts, and then this PR should be good to go.

yashsriv commented:
@YaManicKill Done.

astorije commented:
@yashsriv, I took the liberty to add an extra commit with a few changes. Hope that's OK.
All the details about this commit are in the commit message. With that extra set of changes, I am 👍 if you are OK with my changes.
But now there is a `package.json` conflict 😩 (I know I added a commit, but I didn't want to change your history with a `rebase`!). Alright, couldn't be closer to a go!

I also noted a couple of things that can be improved but not blocking for this PR so I'll open issues for them.

astorije commented:
Ah, stupid question: does this still need `json-loader`?

astorije commented:
One thing I'd love to address though (but I'm sure I'll trigger opinions :p) is that I would prefer to keep only emoji that are very broadly supported. For example, I mentioned `:rofl:` but others as well are not supported in OS X El Capitan (not sure about Sierra).
Maybe keep UTF-9 out of the loop for now? Thoughts?

yashsriv commented:
Rebased 😄 . And yes we do need `json-loader` (for the `require("simplemap.json")` statement)

xPaw commented:
We could do `module.exports` on it, then it will be JavaScript straight away.

MiniDigger commented:
> Maybe keep UTF-9 out of the loop for now? Thoughts?  
that would be Unicode 9, there is not utf8 ^^
and I don't think that we should delay the rollout of Unicode 9 even more, just because one platform is a bit late (as far as I know all others support Unicode 9 by now)

YaManicKill commented:
> and I don't think that we should delay the rollout of Unicode 9 even more

Agreed, we should support them as soon as we can. 

I would even argue we should support Unicode 10 already (don't add to the PR, though, we can add that later) as that is already preliminary released.

YaManicKill commented:
> Rebased :smile: . And yes we do need json-loader (for the require("simplemap.json") statement)

Wait a minute, does webpack not support just including json by default? Sigh...that's pretty standard, I would expect. But yeah, I'd agree with @xPaw

astorije commented:
> just because one platform is a bit late

It's kind of a major platform, and it looks real sloppy on it. But maybe I'm just not up-to-date and it's available on Sierra? (I can't upgrade because of work for reasons)

> Wait a minute, does webpack not support just including json by default?

Actually, [`json-loader`](https://github.com/webpack-contrib/json-loader) says:

> Note: Since webpack v2, JSON files will work by default. You might still want to use this if you use a custom file extension.

When this PR was started, we were under Webpack v1. I'll give it a shot quickly, and also I noticed the following I need to look into (maybe) as there is slight chance I introduced that myself (look at the first line, happens with every search term):

<img width="192" alt="screen shot 2017-04-25 at 22 16 35" src="https://cloud.githubusercontent.com/assets/113730/25406082/d871d93c-2a05-11e7-9e82-6db4cc868267.png">

I also noticed the following, is that expected? It's kinda weird (but we can fix later)

<img width="190" alt="screen shot 2017-04-25 at 22 18 00" src="https://cloud.githubusercontent.com/assets/113730/25406134/fb4eebd4-2a05-11e7-87d2-4f8232cb2b1b.png">

astorije commented:
> Note: Since webpack v2, JSON files will work by default.

Confirmed after testing (I've committed here), this is also [listed in the migrating guide](https://webpack.js.org/guides/migrating/#json-loader-is-not-required-anymore).

astorije commented:
> I noticed the following I need to look into (maybe) as there is slight chance I introduced that myself

It was my fault indeed. It turns out I missed that returning `null` from `$.map` removes an element and:

```js
$.map(emojiMap, function (emoji) { return emoji.indexOf(term) === 0 ? emoji : null; })
```

is the same as:

```js
emojiMap.filter(name => name.indexOf(term) === 0)
```

(and not the `.map` that I initially used)

astorije commented:
I believe I'm all set. @yashsriv, are you okay with my changes?

yashsriv commented:
Yeah. 👍

On Wed 26 Apr, 2017, 2:56 AM Jérémie Astori, <notifications@github.com>
wrote:

> I believe I'm all set. @yashsriv <https://github.com/yashsriv>, are you
> okay with my changes?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/thelounge/lounge/pull/787#issuecomment-297170041>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJ86_L_NCsn6EJIZ7l2wMCu7pyHJZL2wks5rzmUSgaJpZM4LJzGv>
> .
>
-- 

Yash Srivastav
IIT Kanpur CSE
Undergraduate


"There is no fate, but what we make"

xPaw commented on package.json:
```diff
@@ -49,6 +49,7 @@
     "express-handlebars": "3.0.0",
     "fs-extra": "2.1.2",
     "irc-framework": "2.8.0",
+    "jquery-textcomplete": "1.8.0",
```
Shouldn't this be a Dev dep?

astorije replied:
Good catch, yes!
It doesn't need to be a dependency since it gets built into `lounge.vendor.js` by `npm run build` at release time.

@yashsriv, I'm on the spottiest connection ever these days, would you mind making that change?

YaManicKill replied:
I mean, our definition of what is dev dependency and which is dependency is dodgy anyway, as you need the dependency to run the build, and we don't provide the build anywhere, so even running in prod, you'll still need to install the dev dependencies. But whatever makes you happy :-P

xPaw replied:
Prod installations come pre compiled, you don't need dev dependencies.

YaManicKill replied:
Oh yeah, I forgot we pre-built the npm package. Nevermind then :-)

astorije commented:
All, once `jquery-textcomplete` gets moved to `devDependencies`, are we all good with 🚢 this?

YaManicKill commented:
@yashsriv Thanks, there appears to be 1 conflict now that @xPaw merged a different PR. can you do 1 more conflict management and then this'll get merged.

MaxLeiter commented:
🎉🎉🎉🎉
